File: Shortest Path/Dijkstra/22865번 - 가장 먼 곳/22865.cpp
// 22865번 : 가장 먼 곳
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

#define INF 1e9

int n, m;
vector<int> friends;
vector<pair<int, int> > roads[500001];
int min_dist[3][100001] = { 0, };

void input() {
    cin >> n;
    friends.resize(3);
    for(int i = 0; i < 3; i++) {
        cin >> friends[i];
    }
    cin >> m;
    for(int i = 0, d, e, l; i < m; i++) {
        cin >> d >> e >> l;
        roads[d].push_back({e, l});
        roads[e].push_back({d, l});
    }
}

void dijkstra(int idx, int st) {
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>  > pq;
    pq.push({0, st});
    min_dist[idx][st] = 0;
    
    while(!pq.empty()) {
        int dist = pq.top().first;
        int cur = pq.top().second;
        pq.pop();
        
        if(min_dist[idx][cur] < dist) continue;
        
        for(int i = 0; i < roads[cur].size(); i++) {
            int nxt_dist = dist + roads[cur][i].second;
            int nxt = roads[cur][i].first;
            
            if(min_dist[idx][nxt] > nxt_dist) {
                min_dist[idx][nxt] = nxt_dist;
                pq.push({nxt_dist, nxt});
            }
        }
    }
}

void solve() {
    int dist = 0;
    int ans = -1;
    
    fill_n(&min_dist[0][0], 3 * 100001, INF);
    
    // 친구들의 위치에서 각 거리들을 저장
    for(int i = 0; i < 3; i++) {
        dijkstra(i, friends[i]);
    }
    
    for(int i = 1; i <= n; i++) {
        // 친구들의 위치는 건너뜀
        if(i == friends[0] || i == friends[1] || i == friends[2]) continue;
        
        // 친구들의 위치가 아닌 것에서 가장 가까이 갈 수 있는 친구 거리
        int from_away = min(min_dist[0][i], min(min_dist[1][i], min_dist[2][i]));
        
        // 기존 거리보다 더 멀리 존재하는 것이라면 갱신
        if(dist < from_away) {
            dist = from_away;
            ans = i;
        }
    }
    
    cout << ans << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Shortest Path/Dijkstra/22865번 - 가장 먼 곳/22865.java
import java.io.*;
import java.util.*;

public class Main {
    static final int INF = (int) 1e9;
    static int n, m;
    static int[] friends = new int[3];
    static List<int[]>[] roads = new ArrayList[100001];
    static int[][] minDist = new int[3][100001];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        n = Integer.parseInt(br.readLine());

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < 3; i++) {
            friends[i] = Integer.parseInt(st.nextToken());
        }

        m = Integer.parseInt(br.readLine());

        for (int i = 0; i <= n; i++) {
            roads[i] = new ArrayList<>();
        }

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int d = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());
            int l = Integer.parseInt(st.nextToken());

            roads[d].add(new int[]{e, l});
            roads[e].add(new int[]{d, l});
        }

        solve();
    }

    static void dijkstra(int idx, int start) {
        Arrays.fill(minDist[idx], 0, n + 1, INF);

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a[0]));
        pq.offer(new int[]{0, start});
        minDist[idx][start] = 0;

        while (!pq.isEmpty()) {
            int[] cur = pq.poll();
            int dist = cur[0];
            int now = cur[1];

            if (minDist[idx][now] < dist) continue;

            for (int[] next : roads[now]) {
                int to = next[0];
                int cost = dist + next[1];

                if (minDist[idx][to] > cost) {
                    minDist[idx][to] = cost;
                    pq.offer(new int[]{cost, to});
                }
            }
        }
    }

    static void solve() {
        for (int i = 0; i < 3; i++) {
            dijkstra(i, friends[i]);
        }

        int ans = -1;
        int maxDist = 0;

        for (int i = 1; i <= n; i++) {
            if (i == friends[0] || i == friends[1] || i == friends[2]) continue;

            int minFriendDist = Math.min(minDist[0][i], Math.min(minDist[1][i], minDist[2][i]));

            if (minFriendDist > maxDist) {
                maxDist = minFriendDist;
                ans = i;
            }
        }

        System.out.println(ans);
    }
}
